### ISSUE TITLE
[RFC] Having methods for BioSequences types

### ISSUE THREAD

**Issue by camilogarciabotero** [gus-pendleton/CUBScout.jl#8]
Hi @gus-pendleton 

This is such a nice idea to incorporate the metrics of this package! I mentioned on Slack that I'm willing to incorporate some of them in my package [GeneFinder.jl](https://github.com/camilogarciabotero/GeneFinder.jl) in the future. Currently, GeneFinder is able to detect ORFs in a dedicated `struct`, and then using the BioSequences types it can get the `LongSubSeq{DNAAlphabet{4}}`  in a vector of them, here is an example:

```julia
julia> seq = randdnaseq(10^3)

1000nt DNA Sequence:
GGGATTCACATCGTACCGCCCGGGTACCGCCGGCAACGT…TGTCAGACCCATTGAAAGGCCTATGGCATCCTGTGCAAG
```

Using the GeneFinder function `gets()` it extracts all the possible ORFs in a `Vector{LongSubSeq{DNAAlphabet{4}}}`:

```julia
julia> getcds(seq)

36-element Vector{LongSubSeq{DNAAlphabet{4}}}:
 ATGCCATAG
 ATGGGTCTGACACTGCAGTCACTAGTTGCTAATAACAGA…CTGTCTCATATTTCATCAGAAGTGAGACTGTGTTGTTAG
 ATGTACTCAGACGATGGTACACGCAGTCAGACACGAATAGCGGGTAATTAA
 ATGGGAATGAAGCACCAGCAGCGCCTTACAGAAGGCACC…CAAGTGTCCCGAAGCTGTCTCATATTTCATCAGAAGTGA
 ATGAAGCACCAGCAGCGCCTTACAGAAGGCACCTACCGC…CAAGTGTCCCGAAGCTGTCTCATATTTCATCAGAAGTGA
 ATGGTACACGCAGTCAGACACGAATAG
 ATGCTCATGCGACCAGTTAGACCGCAGACTCCGATAGTC…GTACAATCCCTAGTTCGGGGTTCGAGTTCTCCTAGATGA
 ATGCGACCAGTTAGACCGCAGACTCCGATAGTCCGATCG…GTACAATCCCTAGTTCGGGGTTCGAGTTCTCCTAGATGA
 ATGGGGCATTAG
 ATGGGAAACGACAACATATGTACAATCCCTAGTTCGGGGTTCGAGTTCTCCTAG
 ATGTACAATCCCTAG
 ATGATAATTAGGCGTTCACGATACACAATGGAAGGAGAGGCGGGCCCTGCAGGCTAA
 ATGCGTACCTCGTAG
 ATGGAAGGAGAGGCGGGCCCTGCAGGCTAA
 ATGCACAAGTCGCTTGATATGGGGTGGTAG
 ⋮
 ATGCGCTAA
 ATGCTACCGGGTAGGGATAGCCCGATGGGGGGCATGGAC…CAGGGGCTGGACCCCGTTGCAAATAATTCCTCGAAATGA
 ATGGGGGGCATGGACTTAGGCGGCCCCAGCTTATTCTAC…CAGGGGCTGGACCCCGTTGCAAATAATTCCTCGAAATGA
 ATGGACTTAGGCGGCCCCAGCTTATTCTACCACCCCATA…CAGGGGCTGGACCCCGTTGCAAATAATTCCTCGAAATGA
 ATGGTTCTAATCTCGGGCAGCTACGAGGTACGCATGCAA…CAGGGGCTGGACCCCGTTGCAAATAATTCCTCGAAATGA
 ATGCAAAACCACTCCTCGATCTGTCAGGGGCTGGACCCCGTTGCAAATAATTCCTCGAAATGA
 ATGTTGTCGTTTCCCATTTTGTGTCCGATCGGACTATCGGAGTCTGCGGTCTAA
 ATGACTGTACTAGGTCACGGTCAATCCTAA
 ATGCCCCATACGTGA
 ATGAGCATGCGTCGGAAGCTTTGCTTACAACCGCGACAGAGATCAAGCAATTGA
 ATGCGTCGGAAGCTTTGCTTACAACCGCGACAGAGATCAAGCAATTGA
 ATGAAATATGAGACAGCTTCGGGACACTTGGAACCTTAT…CTGCTGGTGCTTCATTCCCATCTAGGAATCTGTTATTAG
 ATGAGACAGCTTCGGGACACTTGGAACCTTATCCTTACAATTGCAATGCGGTAG
 ATGCGGTAG
 ATGTGA
```

So I was curious whether the CUBs or EXPs measurements could be applied to the single CDS or the vector of CDSs. Either way, I was wondering if there is any chance to have `CUBScout` methods that leverage these dedicated data structures as they are optimized for the Bio ecosystem (I meant the BioSequences types by the way).

Best.

**Comment by gus-pendleton** [comment]
Hi Camilo,

GeneFinder is such a cool module, I'm excited to see it! Yes, I think incorporating methods for for BioSequences makes sense. I'll work on it. 

In terms of your use case, the choice of measure will be important depending on the end goal. Some CUB measures, like ENC and SCUO, calculate CUB against a theoretical distribution of codon frequencies, so those can very easily be applied on a single CDS or vector of CDSs. B, ENC', MILC, and MCB, and all the EXPs measurements, calculate codon usage bias of each CDS against a reference set. By default, this is the average codon frequency across the entire set of CDSs. EXP measures (besides GCB) need an additional reference set (often ribosomal genes). I'm not as familiar with ORF-identification, but is it reasonable to use your vector of CDSs from getcds() as a reference set? A key portion of that would be whether a vector of CDSs usually represents ORFs from a single genome. CUB measures are also very sensitive to gene length, and should usually not be applied to genes under at least 80 codons. 

Long story short, for sure you can calculate CUB along a vector of CDSs, and I'll work on writing the methods to do so. But the actual quality of data from CUB will be very dependent on the quality of the CDSs it's being calculated from.

**Comment by camilogarciabotero** [comment]
Hey Gus,

Thanks for answering quickly and explaining some differences. 

Regarding your question: 
>  I'm not as familiar with ORF identification, but is it reasonable to use your vector of CDSs from getcds() as a reference set?

For the time being [GeneFinder.jl](https://github.com/camilogarciabotero/GeneFinder.jl) is only able to find ORFs (currently designed for finding a flexible regex representing an ORF in a DNA sequence). Therefore, depending on the `kwargs` of the function (which tunes for length and alternative codons) the CDS **could also be random or very short** ORF, thus not true CDS. I think, then, that this could be problematic for some of the EXPs metrics, however, this does not necessarily need to affect the design of the EXPs and CUBs functions, but application over the dataset. On the other hand, maybe ENC and SCUO, could have a simpler definition for single CDS (as a `LongSequence` perhaps?) because of the theoretical distribution (How does it work in the package?). What are your thoughts?

Regarding the ORFs prediction (more precisely CDS prediction), other strategies (hopefully will appear in the package) classify and rank those ORFs to reduce the rate of false positives. In fact, one of those is called [ORPHEUS](https://academic.oup.com/nar/article/26/12/2941/2385852?login%3Dtrue) which uses some CUBs metrics to score ORFs.

**Comment by gus-pendleton** [comment]
Oooh I see. It looks like ORPHEUS is calculating their quality metric using a weighted codon frequency, which is data that underlies the CUB measures in CUBScout. That's convenient, since I'm already writing functions to return codon frequencies across a genome and within each gene. That should be useful to you if you're replicating the ORPHEUS approach. 

And for CUB measures for LongSequences, I'll just create methods for ENC and SCUO that accept either single sequences or a vector of sequences - the other measures will only accept a vector of sequences.

**Comment by camilogarciabotero** [comment]
That's awesome! I'll be looking forward to trying those methods!

**Comment by gus-pendleton** [comment]
I've added methods for using BioSequences as input. Because most CUB measures are operating across a genome, they accept Vector{<:NucSeq} as an input. Even though ENC or SCUO could technically accept a single BioSequence and still work, it complicates how they're written and their documentation. As such, if you want to provide a single sequence for ENC or SCUO, you can always wrap it in [].

**Comment by camilogarciabotero** [comment]
Cool! Thanks a lot! I will check that out!
